File: typescript/src/string/case.ts
/**
 * Case conversion implementing the fleet's naming rules (§3.1).
 *
 * Mirrors the Python `capsize_commons.text.case` so a value normalized on one
 * side of the wire matches the other.
 */

const SEPARATORS = /[^0-9A-Za-z]+/g;
const CAMEL_BOUNDARY = /(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])/g;
const COMBINING_MARKS = /[\u0300-\u036f]/g;

function toAscii(value: string): string {
  return value.normalize("NFKD").replace(COMBINING_MARKS, "");
}

function words(value: string): string[] {
  return toAscii(value)
    .trim()
    .replace(CAMEL_BOUNDARY, " ")
    .split(SEPARATORS)
    .filter((word) => word.length > 0);
}

export interface SlugifyOptions {
  /** Joins words; defaults to `"-"`. */
  separator?: string;
  /** Lowercases the result; defaults to `true`. */
  lowercase?: boolean;
}

/** Join the words of `value` with `separator`. */
export function slugify(value: string, options: SlugifyOptions = {}): string {
  const separator = options.separator ?? "-";
  const lowercase = options.lowercase ?? true;
  const parts = words(value);
  const normalized = lowercase
    ? parts.map((word) => word.toLowerCase())
    : parts;
  return normalized.join(separator);
}

/** Return `value` as `kebab-case` (the repo/file convention). */
export function toKebabCase(value: string): string {
  return slugify(value, { separator: "-" });
}

/** Return `value` as `snake_case` (the symbol convention). */
export function toSnakeCase(value: string): string {
  return slugify(value, { separator: "_" });
}

/** Return `value` as `PascalCase` (the class convention). */
export function toPascalCase(value: string): string {
  return words(value)
    .map((word) => {
      const lower = word.toLowerCase();
      return lower.charAt(0).toUpperCase() + lower.slice(1);
    })
    .join("");
}

File: python/src/capsize_commons/text/case.py
"""Case conversion for repository, file and symbol naming (§3.1).

Repos and files are ``kebab-case``, Python symbols ``snake_case`` and classes
``PascalCase``. Every project was writing a slightly different ``slugify``;
this is the one they share.
"""

from __future__ import annotations

import re
import unicodedata

__all__ = ["slugify", "to_kebab_case", "to_pascal_case", "to_snake_case"]

#: Split on any run of characters that are not alphanumerics.
_SEPARATORS = re.compile(r"[^0-9A-Za-z]+")
#: Split camelCase / PascalCase / HTTPServer at the correct boundaries.
_CAMEL_BOUNDARY = re.compile(
    r"(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])"
)


def _to_ascii(text: str) -> str:
    """Fold accents to their closest ASCII spelling."""
    decomposed = unicodedata.normalize("NFKD", text)
    return decomposed.encode("ascii", "ignore").decode("ascii")


def _words(value: str) -> list[str]:
    """Split ``value`` into its alphanumeric words, accents folded."""
    spaced = _CAMEL_BOUNDARY.sub(" ", _to_ascii(value).strip())
    return [word for word in _SEPARATORS.split(spaced) if word]


def slugify(
    value: str, *, separator: str = "-", lowercase: bool = True
) -> str:
    """Join the words of ``value`` with ``separator``.

    >>> slugify("Capsize Persona!")\
    # doctest: +SKIP
    'capsize-persona'
    """
    words = _words(value)
    if lowercase:
        words = [word.lower() for word in words]
    return separator.join(words)


def to_kebab_case(value: str) -> str:
    """Return ``value`` as ``kebab-case`` (the repo/file convention)."""
    return slugify(value, separator="-")


def to_snake_case(value: str) -> str:
    """Return ``value`` as ``snake_case`` (the Python symbol convention)."""
    return slugify(value, separator="_")


def to_pascal_case(value: str) -> str:
    """Return ``value`` as ``PascalCase`` (the class convention)."""
    return "".join(word.capitalize() for word in _words(value))
